File: LeetCode/200. Number of Islands/solution.java
class Solution {
    public int numIslands(char[][] grid) {
        int res = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == '1') {
                    dfsMarking(grid, i, j);
                    res++;
                }
            }
        }
        return res;
    }
    
    private void dfsMarking(char[][] grid, int i, int j) {
        // conditions to stop searching
        if (i < 0 || j < 0 || i >= grid.length || j >= grid[0].length || grid[i][j] == '0') {
            return;
        }
        // mark current island as water or visited, so next time we won't visit it again
        grid[i][j] = '0';
        
        // do dfs to mark all adjacent islands
        dfsMarking(grid, i-1, j);
        dfsMarking(grid, i+1, j);
        dfsMarking(grid, i, j-1);
        dfsMarking(grid, i, j+1);
    }
}

File: LeetCode/200. Number of Islands/solution.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid:
            return 0
        res = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    self.dfs(grid, i, j)
                    res += 1
        return res
    
    def dfs(self, grid: List[List[str]], i: int, j: int):
        if i < 0 or j < 0 or i >= len(grid) or j >= len(grid[0]) or grid[i][j] == '0':
            return
        grid[i][j] = '0'
        
        self.dfs(grid, i-1, j)
        self.dfs(grid, i+1, j)
        self.dfs(grid, i, j-1)
        self.dfs(grid, i, j+1)
